lupa/_lupa.pyx

```diff
@@ -1667,13 +1678,20 @@ cdef _LuaTable py_to_lua_table(LuaRuntime runtime, lua_State* L, items):
     old_top = lua.lua_gettop(L)
     lua.lua_newtable(L)
     # FIXME: how to check for failure?
-
+    if recursive and mapped_objs is None:
+        mapped_objs = {}
     try:
         for obj in items:
+            if recursive:
+                if id(obj) not in mapped_objs:
+                    mapped_objs[id(obj)] = lua.lua_gettop(L)
```

Comment:
Because the call is recursive. If the cache is moved to the end of the loop body, ```py_to_lua``` might trigger a call to  
 ```py_to_lua_table``` again during the loop, which triggers another call and so on, so it might never be able to called.

Reply:
And, this function creates a table at the beginning, which will be filled with data from python objects, and that table will be converted into a ```_LuaTable```, so we just need to keep an eye on that table.

Reply:
Could you try mapping a Python list like `[[3, 3, 3]]` to Lua with this? I.e. one that actually repeats the same value multiple times, so that the cache would be hit? It looks like it might end up referencing the mapped table itself as a value in the Lua table at some point.

Reply:
```[[3, 3, 3]]``` have a depth of 2, it will call ```py_to_lua_table``` twice. Each time the id of the list is cached, and at the second time, after the inner list is cached, it just goes into the list handler of the loop, without caching the PyLongObject. That means, only list and dict is cached.

Reply:
A new test have been added. It works as expected.

Reply:
> And, this function creates a table at the beginning, which will be filled with data from python objects, and that table will be converted into a `_LuaTable`, so we just need to keep an eye on that table.

Ok, so, do I understand correctly that the invariant in this for-loop is that the top of the Lua stack always refers to the Lua table that we are filling? And that is the value that we put into `mapped_objs`?

Why don't you read the stack top once, before the loop, give it a name (like `lua_table_ref`) and use that instead? That would be much clearer than reading seemingly randomly from the top of the Lua stack and putting whatever we get into `mapped_objs` for an object ID that we have never seen before.